### PR TITLE
Add group_callback and label_callback to DC_Folder

### PR DESCRIPTION
**Pull request by markotodic**
Invoke defined callbacks on folder label creation (group_callback)
Invoke defined callbacks on file label/thumbnail creation (label_callback)

### CODE REVIEW

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 1]
```diff
@@ -2662,7 +2663,24 @@ protected function generateTree($path, $intMargin, $mount=false, $blnProtected=t
 
 			// Add the current folder
 			$strFolderNameEncoded = StringUtil::convertEncoding(StringUtil::specialchars(basename($currentFolder)), Config::get('characterSet'));
-			$return .= Image::getHtml($folderImg, '').' <a href="' . $this->addToUrl('fn='.$currentEncoded) . '" title="'.StringUtil::specialchars($GLOBALS['TL_LANG']['MSC']['selectNode']).'"><strong>'.$strFolderNameEncoded.'</strong></a></div> <div class="tl_right">';
+
+            $strFolderLabel = "<strong>$strFolderNameEncoded</strong>";
+
+            // Call the group callback ($strFolderLabel, $objFolder, $protected, $this)
+            if (\is_array($GLOBALS['TL_DCA']['tl_files']['list']['label']['group_callback']))
```
Comment (ausi) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php]: `tl_files` should not be hardcoded here.

```suggestion
            if (\is_array($GLOBALS['TL_DCA'][$this->strTable]['list']['label']['group_callback']))
```

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 2]
```diff
@@ -2597,6 +2597,7 @@ protected function generateTree($path, $intMargin, $mount=false, $blnProtected=t
 			$md5 = substr(md5($folders[$f]), 0, 8);
 			$content = scan($folders[$f]);
 			$currentFolder = StringUtil::stripRootDir($folders[$f]);
+            $objFolder = FilesModel::findByPath($currentFolder);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php]: We should probably check for `$this->blnIsDbAssisted` here.

```suggestion
            $objFolder = $this->blnIsDbAssisted ? FilesModel::findByPath($currentFolder) : null;
```

**Comment by markotodic** [comment]
@leofeyer do I do something about AppVeyor fail or it is up to you to resolve? As you can see I just commited change @ausi suggested.

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 3]
```diff
@@ -2778,6 +2796,20 @@ protected function generateTree($path, $intMargin, $mount=false, $blnProtected=t
...
 			$strFileNameEncoded = StringUtil::convertEncoding(StringUtil::specialchars(basename($currentFile)), Config::get('characterSet'));
 
+            // Call the label_callback ($strFileNameEncoded, $thumbnail, $objFile, $protected, $this)
+            if (\is_array($GLOBALS['TL_DCA'][$this->strTable]['list']['label']['label_callback']))
+            {
+                $strClass = $GLOBALS['TL_DCA'][$this->strTable]['list']['label']['label_callback'][0];
+                $strMethod = $GLOBALS['TL_DCA'][$this->strTable]['list']['label']['label_callback'][1];
+
+                $this->import($strClass);
+                list($strFileNameEncoded, $thumbnail) = $this->$strClass->$strMethod($strFileNameEncoded, $thumbnail, $objFile, $protected, $this);
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php:2806]: I would prefer not to return anything here and modify the values via reference instead.

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 4]
```diff
@@ -2662,7 +2663,24 @@ protected function generateTree($path, $intMargin, $mount=false, $blnProtected=t
 
 			// Add the current folder
 			$strFolderNameEncoded = StringUtil::convertEncoding(StringUtil::specialchars(basename($currentFolder)), Config::get('characterSet'));
-			$return .= Image::getHtml($folderImg, '').' <a href="' . $this->addToUrl('fn='.$currentEncoded) . '" title="'.StringUtil::specialchars($GLOBALS['TL_LANG']['MSC']['selectNode']).'"><strong>'.$strFolderNameEncoded.'</strong></a></div> <div class="tl_right">';
+
+            $strFolderLabel = "<strong>$strFolderNameEncoded</strong>";
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php:2667]:
```suggestion
			$strFolderLabel = '<strong>' . $strFolderNameEncoded . '</strong>';
```

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 5]
```diff
@@ -2597,6 +2597,7 @@ protected function generateTree($path, $intMargin, $mount=false, $blnProtected=t
 			$md5 = substr(md5($folders[$f]), 0, 8);
 			$content = scan($folders[$f]);
 			$currentFolder = StringUtil::stripRootDir($folders[$f]);
+            $objFolder = $this->blnIsDbAssisted ? FilesModel::findByPath($currentFolder) : null;
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php:2600]: It seems you are never using `$objFolder` in this method. Why did you add this line?

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Comment by aschempp** [comment]
@markotodic can you finish this pull request?

**Pull request closed without merging**